src/datatype.c

```diff
@@ -923,6 +924,10 @@ static void init_struct_tail(jl_datatype_t *type, jl_value_t *jv, size_t na)
 JL_DLLEXPORT jl_value_t *jl_new_structv(jl_datatype_t *type, jl_value_t **args, uint32_t na)
 {
     jl_ptls_t ptls = jl_get_ptls_states();
+    if (!jl_is_datatype(type) || type->layout == NULL)
+        jl_type_error("new", (jl_value_t*)jl_datatype_type, (jl_value_t*)type);
+    if ((type->ninitialized > na && !type->mutabl) || na > jl_datatype_nfields(type))
```

Comment:
Since we weren't doing this check at all before it doesn't really matter, but doesn't the `ninitialized` check make sense for both mutable and immutable structs?

Reply:
Hm, yeah, probably. I was concerned about causing too much breakage of otherwise mostly valid code, but it seems that `jl_new_struct_uninit` is usually what's needed when someone wants an incomplete mutable type, so I've gone ahead and removed the extra check here.